### PR TITLE
feat: Priority fee suggestion

### PR DESCRIPTION
**Pull request by dorlneylon**
## Motivation

Removing the constant and instead calculating the capacity in eth api.

## Solution

I'm not yet sure of the realisation, so needs review. For now I'm just getting the lowest value from the cache history, since in itself it should already be filtered by percentiles.

### CODE REVIEW

File: crates/anvil/src/eth/api.rs [hunk 1]
```diff
@@ -1744,7 +1759,7 @@ impl EthApi {
                 base_fee: self.backend.base_fee(),
                 chain_id: self.backend.chain_id().to::<u64>(),
                 gas_limit: self.backend.gas_limit(),
-                gas_price: self.backend.gas_price(),
+                gas_price: self.lowest_suggestion_tip().unwrap_or(1e9 as u128),
```
Comment (mattsse) [inline crates/anvil/src/eth/api.rs:1762]: this should use a constant

File: crates/anvil/src/eth/api.rs [hunk 2]
```diff
@@ -1351,7 +1351,22 @@ impl EthApi {
     /// Handler for ETH RPC call: `eth_maxPriorityFeePerGas`
     pub fn max_priority_fee_per_gas(&self) -> Result<U256> {
         node_info!("eth_maxPriorityFeePerGas");
-        Ok(U256::from(self.backend.max_priority_fee_per_gas()))
+        Ok(U256::from(self.lowest_suggestion_tip().unwrap_or(1e9 as u128)))
+    }
+
+    fn lowest_suggestion_tip(&self) -> Option<u128> {
```
Comment (mattsse) [inline crates/anvil/src/eth/api.rs:1357]: needs docs

File: crates/anvil/src/eth/api.rs [hunk 3]
```diff
@@ -1351,7 +1351,22 @@ impl EthApi {
...
+        let fee_history_cache = self.fee_history_cache.lock();
+        let block_number = self.backend.best_number();
+        let latest_cached_block = fee_history_cache.get(&block_number);
```
Comment (mattsse) [inline crates/anvil/src/eth/api.rs:1360]: this should clone the entry and drop the lock immediately

**Review by mattsse — CHANGES_REQUESTED** [review]
I think this is better than using a constant,
supportive

some suggestions

File: crates/anvil/src/eth/api.rs [hunk 4]
```diff
@@ -1351,7 +1351,26 @@ impl EthApi {
...
-        Ok(U256::from(self.backend.max_priority_fee_per_gas()))
+        Ok(U256::from(self.lowest_suggestion_tip().unwrap_or(1e9 as u128)))
+    }
+
+    /// Returns the suggested fee cap
+    fn lowest_suggestion_tip(&self) -> Option<u128> {
+        let fee_history_cache = {
+            let history = self.fee_history_cache.lock();
+            history.clone()
+        };
```
Comment (DaniPopes) [inline crates/anvil/src/eth/api.rs:1362]:
```suggestion
        let fee_history_cache = self.fee_history_cache.lock();
```

Comment (dorlneylon) [reply crates/anvil/src/eth/api.rs:1362]: I was asked to clone the history and drop the lock, as far as I understood

Comment (DaniPopes) [reply crates/anvil/src/eth/api.rs:1362]: I see, I think that's not necessary.

@mattsse that doesn't seem right, if anything it should be made into a rwlock.

Comment (dorlneylon) [reply crates/anvil/src/eth/api.rs]: done

File: crates/anvil/src/eth/api.rs [hunk 5]
```diff
@@ -2436,7 +2459,7 @@ impl EthApi {
                 m.chain_id = chain_id;
                 m.gas_limit = gas_limit;
                 if gas_price.is_none() {
-                    m.gas_price = self.backend.gas_price();
+                    m.gas_price = self.gas_price().unwrap().to::<u128>();
```
Comment (DaniPopes) [inline crates/anvil/src/eth/api.rs:2462]: We can make `fn gas_price() -> u128`, removing the result and using u128 directly to avoid this

Comment (dorlneylon) [reply crates/anvil/src/eth/api.rs]: I have refactored this, though I hardly doubt this adjustment

**Review by DaniPopes — APPROVED** [review]

**Comment by dorlneylon** [comment]
@mattsse can I have your review, please?

**Comment by dorlneylon** [comment]
@mattsse

**Review by mattsse — APPROVED** [review]

**Pull request merged**